### PR TITLE
Shape Component

### PR DESCRIPTION
**Pull request by InfiniteLee**
Add a `shape` component that when attached to an entity (or child entity) using a `body` with `shape: none` updates the shapes on the CANNON.Body. This enables an easy way to define a `body` made of multiple primitive shapes, which should be a performant alternative to using convex hulls. 

The updating of the body is delayed until `tick` so that if multiple `shape` components are defined, the body is only updated once. Does not currently support removing of shapes. 

Also updated the wireframe code in `body` to handle shape offsets and orientations correctly, as well as multiple shapes per body.

### CODE REVIEW

**Comment by wmurphyrd** [comment]
+1 for this idea
I frequently need a shape different from what three2cannon will produce (e.g. a box that doesn't expand to include child entities or a box that is just a little thicker than the plane geometry it encloses for easier grabbing), and the current process involves writing a custom component each time to listen for the body loaded event and create the shapes

File: src/components/body/body.js [hunk 1]
```diff
@@ -71,25 +70,104 @@ var Body = {
...
       this._play();
     }
 
-    this.el.emit('body-loaded', {body: this.el.body});
+    if (this.isLoaded) {
+      this.el.emit('body-loaded', {body: this.el.body});
+    }
+  },
+
+  addShape(shapeData) {
```
Comment (donmccurdy) [inline src/components/body/body.js]: not using ES6 syntax here yet, so this must be:

```js
addShape: function (shapeData) {
```

File: src/components/body/body.js [hunk 2]
```diff
@@ -71,25 +70,104 @@ var Body = {
...
+    if (this.data.shape !== 'none') {
+      console.warn('shape can only be added if shape property is none');
+      return;
+    }
+
+    var type = shapeData.shape;
+    var scale = this.el.object3D.scale;
```
Comment (donmccurdy) [inline src/components/body/body.js]: Let's get world-space scale:

```js
var scale = new THREE.Vector3();
this.el.object3D.getWorldScale( scale );
```

File: src/components/body/body.js [hunk 3]
```diff
@@ -173,30 +251,44 @@ var Body = {
...
+
+      if (orientation) {
+        orientation.inverse(orientation);
+
+        wireframe.quaternion.set(
+          orientation.x,
+          orientation.y,
+          orientation.z,
+          orientation.w
+        );
```
Comment (donmccurdy) [inline src/components/body/body.js]: I think `wireframe.quaternion.copy(orientation)` should work here, right? unless there was some hitch between the cannonjs and threejs quaternion...

File: src/components/body/body.js [hunk 4]
```diff
@@ -173,30 +251,44 @@ var Body = {
...
+        new THREE.LineBasicMaterial({color: 0xff0000})
       );
-    }
 
+      if (offset) {
+        wireframe.position.copy(offset);
+      }
+
+      if (orientation) {
+        orientation.inverse(orientation);
```
Comment (donmccurdy) [inline src/components/body/body.js]: This is a reference to a quaternion still in `body.shapeOrientations`, so I think we want to make a copy before inverting.

File: src/components/shape/shape.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  schema: {
+    shape: {default: 'box', oneOf: ['box', 'sphere', 'cylinder']},
+    radius: {type: 'number', default: 1},
+    offset: {type: 'vec3', default: {x: 0, y: 0, z: 0}},
+    halfExtents: {type: 'vec3', default: {x: 1, y: 1, z: 1}},
+    orientation: {type: 'vec4', default: {x: 0, y: 0, z: 0, w: 1}},
+    radiusTop: {type: 'number', default: 1},
+    radiusBottom: {type: 'number', default: 1},
+    height: {type: 'number', default: 1},
+    numSegments: {type: 'int', default: 8}
```
Comment (donmccurdy) [inline src/components/shape/shape.js]: let's prefix these properties:

```
sphereRadius
boxHalfExtents
cylinderRadiusTop
...
```

File: src/components/shape/shape.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  },
+
+  _findType: function(el) {
+    if (el.components.hasOwnProperty('body')) {
+      return 'body';
+    } else if (el.components.hasOwnProperty('dynamic-body')) {
+      return 'dynamic-body';
+    } else if (el.components.hasOwnProperty('static-body')) {
+      return'static-body';
+    }
```
Comment (donmccurdy) [inline src/components/shape/shape.js]: Would do `el.hasAttribute('foo-body')` here instead

File: src/components/shape/shape.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+  initShape: function() {
+    this.bodyEl = this.el;
+    var type = this._findType(this.bodyEl);
+
+    while (!type && this.bodyEl.parentNode) {
+      this.bodyEl = this.bodyEl.parentNode;
+      type = this._findType(this.bodyEl);
+    }
+    this.bodyEl.components[type].addShape(this.data);
```
Comment (donmccurdy) [inline src/components/shape/shape.js]: What do you think about having the `shape` component construct a `CANNON.Shape` instance and pass that into components.body.addShape( shape )`?

File: src/components/shape/shape.js [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    } else if (el.components.hasOwnProperty('static-body')) {
+      return'static-body';
+    }
+    return null;
+  },
+
+  remove: function() {
+    if (this.bodyEl.parentNode) {
+      console.warn('removing shape component not currently supported');
+    }
```
Comment (donmccurdy) [inline src/components/shape/shape.js]: Is there a case where `remove` could be called but no parent node exists?

Comment (InfiniteLee) [reply src/components/shape/shape.js]: If the parent is removed first we don't need to warn about the shape being removed.

File: src/components/shape/shape.js (hunk 6 above)
Comment (InfiniteLee) [reply src/components/shape/shape.js]: I'm not familiar with this syntax, is `foo-` a wildcard match or something?

File: src/components/body/body.js (hunk 3 above)
Comment (InfiniteLee) [reply src/components/body/body.js]: Is that true of going between cannonjs and threejs Vector3's as well?

File: src/components/body/body.js (hunk 4 above)
Comment (InfiniteLee) [reply src/components/body/body.js]: will add .clone() where `orientation` is set. 
```
orientation = this.body.shapeOrientations[i].clone(),
```

File: src/components/shape/shape.js (hunk 5 above)
Comment (InfiniteLee) [reply src/components/shape/shape.js]: I would prefer not to- I really like that this exactly matches what those properties are called in threejs. (Selfishly I am also writing some tooling upstream of this that I don't want to have to constantly try to remember what the properties are supposed to be called). I would argue that if we want to be more explicit, we should add the other shape components back in (`sphere-shape`, `box-shape`, `cylinder-shape`)

File: src/components/body/body.js (hunk 3 above)
Comment (InfiniteLee) [reply src/components/body/body.js]: Seems to be the case.

File: src/components/body/body.js (hunk 4 above)
Comment (InfiniteLee) [reply src/components/body/body.js]: oh nevermind that's a threejs method, not a cannonjs one.

File: src/components/shape/shape.js (hunk 7 above)
Comment (InfiniteLee) [reply src/components/shape/shape.js]: i have to think about that one- I vaguely recall I had a reason not to do that, but that might not be the case anymore.

File: src/components/shape/shape.js (hunk 5 above)
Comment (wmurphyrd) [reply src/components/shape/shape.js]: What about adding a `updateSchema` handler a la `geometry` that makes the right properties available for the selected shape? That may provide the same clarity of purpose as prefixing

Comment (donmccurdy) [reply src/components/shape/shape.js]: Ok, I'm fine with keeping the names then. But even easier than `updateSchema`, we can do this:

```
radiusTop: { type: 'number', default: 1, if: {shape: ['cylinder']} }
...
```

File: src/components/shape/shape.js (hunk 7 above)
Comment (donmccurdy) [reply src/components/shape/shape.js]: Ok, feels cleaner IMO but if there's some issue we can leave it alone for now.

**Comment by donmccurdy** [comment]
Nice work, thanks! Will test this out a bit and then cut a new release.

**Comment by donmccurdy** [comment]
Published 3.1.0. 🎊

**Comment by jewingo** [comment]
It looks like compound colliders break when an orientation property is specified (3.3.0) for any component shape; an alternative is using the constraint functionality to create a complex physics body (rotation seems to work fine for primitive body shapes) but I'm worried about overhead with this methodology. Looking into using [wmurphyrd's aframe-physics-extras](https://github.com/wmurphyrd/aframe-physics-extras) for this

**Pull request merged**